Docs for request logging (#12363)
apache/druid: docs/operations/request-logging.md

---
id: request-logging
title: Request logging
sidebar_label: Request logging
---

<!--
  ~ Licensed to the Apache Software Foundation (ASF) under one
  ~ or more contributor license agreements.  See the NOTICE file
  ~ distributed with this work for additional information
  ~ regarding copyright ownership.  The ASF licenses this file
  ~ to you under the Apache License, Version 2.0 (the
  ~ "License"); you may not use this file except in compliance
  ~ with the License.  You may obtain a copy of the License at
  ~
  ~   http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing,
  ~ software distributed under the License is distributed on an
  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  ~ KIND, either express or implied.  See the License for the
  ~ specific language governing permissions and limitations
  ~ under the License.
  -->

> All Apache Druid services that can serve queries can also log the query requests they see.

Reviewer: ```suggestion
All Apache Druid services that can serve queries can also log the query requests they process.
```
Not sure how sight figures in.

Request logs contain information on query metrics, including execution time and memory usage.
You can use information in the request logs to monitor query performance, determine bottlenecks, and analyze and improve slow queries.

Request logging is disabled by default.

> This topic describes how to configure Druid to log and emit query metrics.

Reviewer 2: Is this query metrics? Its more of query logs with associated metrics and dimensions
@suneet-s

Author: The logs contain the metrics like [`sqlQuery/time`](https://druid.apache.org/docs/latest/operations/metrics.html#sql-metrics) so not sure if this is a matter of semantics. 🤷 
If it's clearer, we could change the phrasing to be something like the following:
```suggestion
This topic describes how to configure Druid to generate request logs to track query metrics.
```

Reviewer 2: thats better

- [`slf4j`](../configuration/index.md#slf4j-request-logging): Logs queries via the SLF4J Java logging API.
- [`filtered`](../configuration/index.md#filtered-request-logging): Filters requests by query type or execution time before logging the filtered queries by the delegated request logger.
- [`composing`](../configuration/index.md#composing-request-logging): Logs all requests to multiple request loggers.
- [`switching`](../configuration/index.md#switching-request-logging): Logs native queries and SQL queries to separate request loggers.

Define the type of request logger in `druid.request.logging.type`.
See the [Request logging configuration](../configuration/index.md#request-logging) for properties to set for each type of request logger.
Specify these properties in the `common.runtime.properties` file.
You must restart Druid for the changes to take effect.

Druid stores the results in the Broker logs, unless the request logging type is `emitter`.
If you use emitter request logging, you must also configure metrics emission.

## Configure metrics emission

> Druid emits metrics and alerts via an emitter. 

Reviewer: ```suggestion
Druid includes various emitters to send metrics and alerts. 
```

- [`composing`](../configuration/index.md#composing-emitter-module): Emits metrics to multiple emitter types.
- [`graphite`](../configuration/index.md#graphite-emitter): Emits metrics to a [Graphite](https://graphiteapp.org/) Carbon service.

Specify these properties in the `common.runtime.properties` file.
See the [Metrics emitters configuration](../configuration/index.md#metrics-emitters) for properties to set for each type of metrics emitter.
You must restart Druid for the changes to take effect.

## Example

The following configuration shows how to enable request logging and post query metrics to the endpoint `http://example.com:8080/path`.
```
# Enable request logging and configure the emitter request logger
druid.request.logging.type=emitter
druid.request.logging.feed=myRequestLogFeed

# Enable metrics emission and tell Druid where to emit messages
druid.emitter=http
druid.emitter.http.recipientBaseUrl=http://example.com:8080/path

# Authenticate to the base URL, if needed
druid.emitter.http.basicAuthentication=username:password
```

> The following output shows an example of the emitted information:

Reviewer: ```suggestion
The following shows an example log emitter output:
```
